src/psyclone/psyir/tools/dependency_tools.py

```diff
@@ -374,3 +374,83 @@ def can_loop_be_parallelised(self, loop, loop_variable=None,
                     return False
 
         return result
+
+    # -------------------------------------------------------------------------
+    def get_input_parameters(self, node_list, variables_info=None):
+        # pylint: disable=no-self-use
+        '''Return all variables that are input parameters, i.e. are
+        read (before potentially being written).
+
+        :param node_list: list of PSyIR nodes to be analysed.
+        :type loop:list of :py:class:`psyclone.psyGen.Node`
+        :param variables_info: optional variable usage information, \
+            can be used to avoid repeatedly collecting this information.
+        :type variables_info: \
+            :py:class:`psyclone.core.variables_info.VariablesAccessInfo`
+
+        :returns: a list of all variable names that are read.
+        :rtype: list of str
+
+        '''
+        # Collect the information about all variables used:
+        if not variables_info:
+            variables_info = VariablesAccessInfo(node_list)
+
+        input_list = []
+        for var_name in variables_info.all_vars:
+            # Take the first access (index 0) of this variable. Note that
+            # loop variables have a WRITE before a READ access, so they
+            # will be ignored
+            first_access = variables_info[var_name][0]
+            # If the first access is a write, the variable is not an input
+            # parameters and does not need to be saved.
+            if first_access.access_type != AccessType.WRITE:
+                input_list.append(var_name)
+
+        return input_list
+
+    # -------------------------------------------------------------------------
+    def get_output_parameters(self, node_list, variables_info=None):
+        # pylint: disable=no-self-use
+        '''Return all variables that are output parameters, i.e. are
+        written.
+
+        :param node_list: list of PSyIR nodes to be analysed.
+        :type loop:list of :py:class:`psyclone.psyGen.Node`
+        :param variables_info: optional variable usage information, \
+            can be used to avoid repeatedly collecting this information.
+        :type variables_info: \
+            :py:class:`psyclone.core.variables_info.VariablesAccessInfo`
+
+        :returns: a list of all variable names that are written.
+        :rtype: list of str
+
+        '''
+        # Collect the information about all variables used:
+        if not variables_info:
+            variables_info = VariablesAccessInfo(node_list)
+
+        output_list = []
+        for var_name in variables_info.all_vars:
+            if variables_info.is_written(var_name):
+                output_list.append(var_name)
+
+        return output_list
+
+    # -------------------------------------------------------------------------
+    def get_in_out_parameters(self, node_list):
+        '''Return a pair of lists that contains all variables that are input
```

Comment:
s/pair/2-tuple/ and s/contains/contain/

Reply:
The first part is done. I am not sure is the second part is correct: "A pair" is singular, i.e. it should be contains. If each of the two lists in the pair would contain input- and output-variables, then it should be contain, but since it is only the pair that contains both type of variables, I _think_ contains should stay ... or?